File: project RAD/anagaram problem using hashmap/sol.java
import java.util.*;

public class sol {

    public static List<List<String>> groupAnagrams(String[] strs) {
        // Step 1: Create a HashMap to store key (sorted word) → list of words
        Map<String, List<String>> map = new HashMap<>();

        for (String word : strs) {
            // Step 2: Convert word to char array, sort it, and make a key
            char[] chars = word.toCharArray();
            Arrays.sort(chars);
            String key = new String(chars);

            // Step 3: Add the original word into the correct group in the map
            map.computeIfAbsent(key, k -> new ArrayList<>()).add(word);
        }

        // Step 4: Return all grouped lists
        return new ArrayList<>(map.values());
    }

    public static void main(String[] args) {
        // --- Test Cases ---
        String[][] testCases = {
                { "eat", "tea", "tan", "ate", "nat", "bat" },
                { "", "" },
                { "a" },
                { "abc", "bca", "cab", "xyz" },
                { "listen", "silent", "enlist", "hello" }
        };

        int[] expectedGroups = { 3, 1, 1, 2, 2 };

        System.out.println("--- Group Anagrams Test Results ---");

        for (int i = 0; i < testCases.length; i++) {
            List<List<String>> result = groupAnagrams(testCases[i]);
            int groupCount = result.size();

            String status = (groupCount == expectedGroups[i]) ? "PASSED"
                    : "FAILED (Expected " + expectedGroups[i] + ", Got " + groupCount + ")";

            // For clean display, sort each group and then the outer list
            for (List<String> group : result) {
                Collections.sort(group);
            }
            result.sort(Comparator.comparing(o -> o.get(0)));

            System.out.println("Input: " + Arrays.toString(testCases[i]));
            System.out.println("Groups: " + result);
            System.out.println("Status: " + status + "\n");
        }
    }
}

File: project RAD/anagaram problem using hashmap/sol.py
from typing import List, Dict
from collections import defaultdict

def groupAnagrams(strs: List[str]) -> List[List[str]]:
    """
    Groups anagrams using a Hash Map where the key is the sorted version 
    of the word (the canonical key).
    
    Time Complexity: O(N * K log K), where N is the number of strings and 
                     K is the average length of a string (due to sorting each string).
    Space Complexity: O(N * K), for storing the map and the resulting groups.
    """
    # Use defaultdict to automatically initialize a list for a new key
    # Key: Sorted string (e.g., "aet"), Value: List of original strings
    anagram_groups: Dict[str, List[str]] = defaultdict(list)
    
    for word in strs:
        # 1. Create the Canonical Key: Sort the characters and join them back into a string
        # Example: "eat" -> ['e', 'a', 't'] -> ['a', 'e', 't'] -> "aet"
        key = "".join(sorted(word))
        
        # 2. Use the key to group the original word in the Hash Map
        anagram_groups[key].append(word)
        
    # 3. Return all the grouped lists (the values of the Hash Map)
    return list(anagram_groups.values())

# --- Test Cases ---

if __name__ == "__main__":
    test_cases = [
        (["eat", "tea", "tan", "ate", "nat", "bat"], 3),
        (["", ""], 1),
        (["a"], 1),
        (["abc", "bca", "cab", "xyz"], 2),
        (["listen", "silent", "enlist", "hello"], 2),
    ]

    print("--- Group Anagrams Test Results ---")
    
    for strs, expected_group_count in test_cases:
        result = groupAnagrams(strs)
        result_group_count = len(result)
        
        # We only check the number of groups, as the internal order can vary
        status = "PASSED" if result_group_count == expected_group_count else f"FAILED (Expected {expected_group_count} groups, Got {result_group_count})"
        
        # To make the output clean, we sort the inner lists and the outer list
        # This is for display verification only, not part of the algorithm logic
        display_result = sorted([sorted(group) for group in result])
        
        print(f"Input: {strs}\nGroups: {display_result}, Status: {status}\n")